src/internal/hooks/use-modal-component-analytics/index.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useEffect } from 'react';
+
+import { PerformanceMetrics } from '../../analytics';
+import { ModalContextProps, useModalContext } from '../../context/modal-context';
+import { useEffectOnUpdate } from '../use-effect-on-update';
+
+export const usePrimaryButtonModalComponentAnalytics = (
+  isPrimaryButton: boolean,
+  elementRef: React.RefObject<HTMLElement>,
+  dependencies: React.DependencyList
+) => {
+  const modalContext = useModalContext();
+  useEffect(() => {
+    if (!isPrimaryButton || !elementRef.current || !modalContext.isInModal) {
+      return;
+    }
+    modalContext.componentLoadingCount.current++;
+    if (!isElementVisible(elementRef.current)) {
+      return;
+    }
+
+    modalContext.componentLoadingCount.current--;
+    emitModalContentReadyMetric(modalContext);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const isElementVisible = (htmlElement: HTMLElement) => {
+    if (!htmlElement) {
+      return true;
+    }
+    return (
+      htmlElement.offsetWidth > 0 &&
+      htmlElement.offsetHeight > 0 &&
+      getComputedStyle(htmlElement).visibility !== 'hidden'
+    );
+  };
+
+  useEffectOnUpdate(() => {
+    if (!isPrimaryButton || !elementRef.current || !modalContext.isInModal) {
+      return;
+    }
+    const elementVisible =
+      elementRef.current.offsetWidth > 0 &&
+      elementRef.current.offsetHeight > 0 &&
+      getComputedStyle(elementRef.current).visibility !== 'hidden';
+
+    if (!elementVisible) {
+      return;
+    }
+
+    modalContext.componentLoadingCount.current--;
+    emitModalContentReadyMetric(modalContext);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, dependencies);
+};
+
+export const useModalComponentAnalytics = () => {
```

Comment:
To me, this function name doesn't make it entirely clear that it reports the current component as a loading component. Maybe we could make this a bit clearer with a more explicit name.

A suggestion:
```suggestion
export const useModalContextLoadingComponent = () => {
```

Reply:
will update